agent/dns.go

```diff
@@ -1760,56 +1760,79 @@ func (d *DNSServer) handleRecurse(resp dns.ResponseWriter, req *dns.Msg) {
 	if _, ok := resp.RemoteAddr().(*net.TCPAddr); ok {
 		network = "tcp"
 	}
+	rejectRecurse := false
 
-	// Recursively resolve
-	c := &dns.Client{Net: network, Timeout: cfg.RecursorTimeout}
-	var r *dns.Msg
-	var rtt time.Duration
-	var err error
-	for _, recursor := range cfg.Recursors {
-		r, rtt, err = c.Exchange(req, recursor)
-		// Check if the response is valid and has the desired Response code
-		if r != nil && (r.Rcode != dns.RcodeSuccess && r.Rcode != dns.RcodeNameError) {
-			d.logger.Debug("recurse failed for question",
-				"question", q,
-				"rtt", rtt,
-				"recursor", recursor,
-				"rcode", dns.RcodeToString[r.Rcode],
-			)
-			// If we still have recursors to forward the query to,
-			// we move forward onto the next one else the loop ends
-			continue
-		} else if err == nil || (r != nil && r.Truncated) {
-			// Compress the response; we don't know if the incoming
-			// response was compressed or not, so by not compressing
-			// we might generate an invalid packet on the way out.
-			r.Compress = !cfg.DisableCompression
-
-			// Forward the response
-			d.logger.Debug("recurse succeeded for question",
-				"question", q,
-				"rtt", rtt,
-				"recursor", recursor,
-			)
-			if err := resp.WriteMsg(r); err != nil {
-				d.logger.Warn("failed to respond", "error", err)
-			}
-			return
+	// Refuse to recursively resolve requests of type DS for domains that we manage.
+	// Upstream resolvers will not know about the correct record anyways, and it might result in a resolver loop
+	if q.Qtype == dns.TypeDS {
```

Comment:
Could you extract a func for this block? Eg `func (d *DNSServer) refuseRecurseDS(req *dns.Msg, domain string) bool`.

And also a func for creating the message `func generateMessage(code int)`.

Afterwards we can write:

```golang
for _, domain := range []string{d.domain, d.altDomain} {
  if d.refuseRecurseDS(req, domain) {
    d.logger.Debug("request for %v (%s): Refusing to recurse type DS requests to %v.", q, network, domain)
    m := generateMessage(dns.RcodeRefused)
    resp.WriteMsg(m)
    return
  }
}
```

That would also get rid of the big indentation below. And `generateMessage` could be reused for the `dns.RcodeServerFailure` message at the end.